Added support for FHIR R4 4.0.1

AndrewBird81 opened the pull request:
Also updated version from 1.3.1 -> 1.3.2 and bumped lodash version.

I tried to mimic the existing code, so there is duplication in the unit test. The most noticeable difference between 4.0.0 and 4.0.1 is the structure of the `Extension` type. In the 4.0.0 modelinfo.xml, the `url` element is defined as:
'<ns4:element` name="url" elementType="System.String"/>`

But in 4.0.1, we have:
`<ns4:element name="url" elementType="FHIR.uri"/>`

As a consequence, an `Extension`'s `url` is no longer a simple string, but a complex object with its own `value`, `id`, `url`, and `Extension` array.

Commit:
Added support for FHIR 4.0.1 R4
Updated version from 1.3.1 -> 1.3.2
Bumped lodash version

brynrhodes commented:
So, this is interesting, we had to make a specific exception for the URL attribute in the ModelInfo generation logic because the url element in HAPI is actually a [URIType](https://github.com/hapifhir/org.hl7.fhir.core/blob/master/org.hl7.fhir.r4/src/main/java/org/hl7/fhir/r4/model/Extension.java#L102)

We could address this in the model info and update the Java provider to account for the discrepancy in the Data Access layer. Thoughts?

AndrewBird81 commented:
Apologies @brynrhodes , but I'm not understanding where the issue lies. Are you referring to the mismatch between the name `url` and the type `uri`? Or are you saying the `url` field in FHIR Extensions should be a System.String, as it was in the 4.0.0 modelinfo.xml? Or is it something else entirely that I'm missing?

hossenlopp commented on test/r401_test.js:
```diff
@@ -0,0 +1,336 @@
+const patientMyron = require('./fixtures/r4/Luna60_McCullough561_6662f0ca-b617-4e02-8f55-7275e9f49aa0.json');
+const patientShawnee = require('./fixtures/r4/Johnnie679_Hermiston71_2cd30bd6-3a87-4191-af90-6daa70f58f55.json');
+
+describe('#FHIRWrapper_R4', () => {
```
Update this describe be different than the 4.0.0 tests. In the current state it's a bit confusing to know which test results are for which FHIR version.

AndrewBird81 replied:
Good point- I updated both descriptions to avoid confusion.

AndrewBird81 replied:
I also removed the unused test fixture. As for the changes to the other unit tests, they were importing the same files multiple times, so I refactored them to be more efficient. I've grown accustomed to the ES6 and TypeScript modules, so seeing dynamic requires has begun to look like an anti-pattern to me. But I could revert them if you prefer.

hossenlopp replied:
I think that makes sense. But a possible downside is that if a test ends up being destructive to the data loaded it could affect other tests. I think that could have been why it was done in the `beforeEach`. I'll leave this to @cmoesel if he thinks this refactoring is ok.

AndrewBird81 replied:
For what it's worth, they're already modified in the `before()` in a safe way:
`conditionResourceWithNoType = JSON.parse(JSON.stringify(conditionResource));`
`delete conditionResourceWithNoType.resourceType;`

hossenlopp reviewed (changes requested):
👍 ModelInfo file matches the version at https://github.com/cqframework/clinical_quality_language/blob/master/Src/java/quick/src/main/resources/org/hl7/fhir/fhir-modelinfo-4.0.1.xml

EXM104 Bundle file fixture is not used by any tests. Perhaps remove it.

Why are there changes in the DSTU2, STU3 and R4 test files?

brynrhodes commented:
@AndrewBird81 , sorry for the confusion, yes, the 4.0.1 model info is correct based on the published spec, I thought we had this workaround, but the workaround was actually for 4.0.0 and it was corrected in 4.0.1. Again, sorry for the confusion, disregard. :)

Commit:
Removing unused test fixture and updating unit test descriptions to avoid confusion.

cmoesel commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "cql-exec-fhir",
-  "version": "1.3.1",
+  "version": "1.3.2",
```
Since this adds support for a new model (FHIR 4.0.1), I'd consider that a _minor_ release rather than a _patch_ release; it now supports entire CQL libraries it did not support before.  So I'd suggest we bump to 1.4.0.  Would you agree?

cmoesel commented on test/r401_test.js:
```diff
@@ -0,0 +1,336 @@
+
+const conditionResource = require('./fixtures/r4/Condition_f201.json');
+const patientMyron = require('./fixtures/r4/Luna60_McCullough561_6662f0ca-b617-4e02-8f55-7275e9f49aa0.json');
+const patientShawnee = require('./fixtures/r4/Johnnie679_Hermiston71_2cd30bd6-3a87-4191-af90-6daa70f58f55.json');
```
These patients are actually Luna and Johnnie, so we should rename the variables to `patientLuna` and `patientJohnnie`.

cmoesel commented on test/r4_test.js:
```diff
@@ -2,14 +2,16 @@ const cql = require('cql-execution');
-describe('#FHIRWrapper_R4', () => {
+const conditionResource = require('./fixtures/r4/Condition_f201.json');
+const patientMyron = require('./fixtures/r4/Luna60_McCullough561_6662f0ca-b617-4e02-8f55-7275e9f49aa0.json');
+const patientShawnee = require('./fixtures/r4/Johnnie679_Hermiston71_2cd30bd6-3a87-4191-af90-6daa70f58f55.json');
```
These patients are actually Luna and Johnnie, so we should rename the variables to `patientLuna` and `patientJohnnie`.

cmoesel commented on test/r401_test.js:
```diff
@@ -0,0 +1,336 @@
+  });
+});
+
+describe('#R4', () => {
```
Can you update this to:
```js
describe('#R4 v4.0.1', () => {
```
like you did for the FHIRWrapper test above?  That way it's easier to distinguish failing 4.0.1 tests from failing 4.0.0 tests.

cmoesel commented on test/r401_test.js:
```diff
@@ -0,0 +1,336 @@
+
+  it('should support getting an option of a choice using explicit name', () => {
+    // This was needed because the ModelInfo used to indicate MedicationRequest's primaryCodePath as medicationCodeableConcept!
+    // Keep it in case anyone tries to load an older ModelInfo that still does that.
```
I don't think the 4.0.1 modelinfo has this legacy, so it's not important for 4.0.1 to support this.  We can probably remove this test.

cmoesel commented on test/r401_test.js:
```diff
@@ -0,0 +1,336 @@
+  it('should support getting an option of a choice', () => {
+    const pt = patientSource.currentPatient();
+    const condition = pt.findRecords('Condition').find(p => p.getId() === '9934bc4f-58af-4ecf-bb70-b7cc31987fc5');
+    // In STU3, you use the stub of the choice (e.g., onset[x] datetime is retrieved as onset)
```
Oops.  This should probably say "R4" instead of "STU3".  I'm guessing this same typo existing in the 4.0.0 test suite as well (and that you merely copy/pasted it).

cmoesel commented on test/r401_test.js:
```diff
@@ -0,0 +1,336 @@
+  });
+
+  it('should not return wrong type if explicit choice name was requested but data used different type', () => {
+    // This is needed because the ModelInfo indicates MedicationRequest's primaryCodePath as medicationCodeableConcept!
```
Similarly, the 4.0.1 modelinfo has `primaryCodePath` `medication`, so I don't think this test is relevant for 4.0.1.

cmoesel commented on test/r4_test.js:
```diff
@@ -2,14 +2,16 @@ const cql = require('cql-execution');
+const patientMyron = require('./fixtures/r4/Luna60_McCullough561_6662f0ca-b617-4e02-8f55-7275e9f49aa0.json');
+const patientShawnee = require('./fixtures/r4/Johnnie679_Hermiston71_2cd30bd6-3a87-4191-af90-6daa70f58f55.json');
+
+describe('#FHIRWrapper_R4 v4.0.0', () => {
```
Adding `v4.0.0` to the suite name is good.  Can you do the same for the other `describe` in this file?

cmoesel reviewed (changes requested):
Thanks for this, @AndrewBird81!  It will be great to have support for FHIR 4.0.1!  I've requested a few small changes -- nothing major at all.

As for the slight refactor in the existing tests, I'm fine with eliminating the repeated calls to `require` in the `beforeEach` clauses.  I suspect that maybe at one point we were modifying the definitions?  I don't know -- but I'm fine w/ the change.  And if the _production_ code mutated the patient data, I'd want to know about that!  So I guess that in a weird way this ends up testing for that too?

Commit:
Addressing PR feedback:
- Changed version to 1.4.0
- Fixed variable names/comments
- Removed unnecessary tests

AndrewBird81 commented:
@cmoesel - excellent feedback! I believe I've made the requested changes. Kindly let me know if you find anything else. Thanks again!

cmoesel commented:
Thanks, @AndrewBird81.  I'd like to manually test this w/ some CQL and R4 patients later today before I officially approve.

cmoesel reviewed (approved):
This looks great!  I've run it manually with some FHIR 4.0.1 CQL and also confirmed it there.  Thanks for the contribution, @AndrewBird81!

hossenlopp reviewed (approved):
Thanks for doing this work, @AndrewBird81!

cmoesel commented:
This has now been released to NPM as version 1.4.0.

GitHub: https://github.com/cqframework/cql-exec-fhir/releases/tag/v1.4.0
NPM: https://www.npmjs.com/package/cql-exec-fhir/v/1.4.0
